src/core/device.c

```
@@ -132,6 +299,18 @@ static int device_load(Unit *u) {
                         log_unit_debug_errno(u, r, "Failed to unescape name: %m");
         }
 
+        root_slice_unit = manager_get_unit(u->manager, SPECIAL_ROOT_SLICE);
```

Maybe it would be better to check `verify_io_cost_supported()` here since it's required for both of these settings? Also I think typically we log, ignore the error, and return early if something is unsupported instead of failing hard. Something like:
```
if (!verify_io_cost_supported()) {
  // log and return message about being not supported but don't fail hard
}

// check for root slice
// call root_cgroup_apply_block_device_io_cost_qos and root_cgroup_apply_block_device_io_cost_model
```

---

Another remark related to this line -- while even the systemd user instance supports the device units, it won't work within its user `-.slice` (and the root in a container isn't the host root neither, iocost config won't apply there).
All in all, `manager_owns_host_root_cgroup()` should be additionally checked too.

---

I agree with werkov on this point you should also check `manager_owns_host_root_cgroup()`. Good candidate for adding to `verify_io_cost_supported()`